### PR TITLE
pipeline send commands to all nodes and then read sequentially

### PR DESCRIPTION
**Pull request by 72squared**
addresses issues #97 and #98

### CODE REVIEW

**Comment by Grokzen** [comment]
@72squared I do think that this would also improve the situation for #99 because of the new way pipelines works. I would consider closing that issue also when this is merged.

File: rediscluster/client.py [hunk 1]
```diff
@@ -96,7 +96,7 @@ class StrictRedisCluster(StrictRedis):
     )
 
     def __init__(self, host=None, port=None, startup_nodes=None, max_connections=32, init_slot_cache=True,
-                 pipeline_use_threads=True, readonly_mode=False, reinitialize_steps=None, **kwargs):
```
Comment (Grokzen) [inline rediscluster/client.py]: This change would require a note in the `docs/Upgrading.md` file stating that this argument can be removed from the implementers code.

File: rediscluster/client.py [hunk 2]
```diff
@@ -154,7 +151,7 @@ def __repr__(self):
     def pubsub(self, **kwargs):
         return ClusterPubSub(self.connection_pool, **kwargs)
 
-    def pipeline(self, transaction=None, shard_hint=None, use_threads=None):
```
Comment (Grokzen) [inline rediscluster/client.py]: This change would require a note in the docs/Upgrading.md file stating that this argument can be removed from the implementers code.

File: rediscluster/client.py [hunk 3]
```diff
@@ -850,7 +846,7 @@ class RedisCluster(StrictRedisCluster):
         }
     )
 
-    def pipeline(self, transaction=True, shard_hint=None, use_threads=None):
```
Comment (Grokzen) [inline rediscluster/client.py]: This change would require a note in the docs/Upgrading.md file stating that this argument can be removed from the implementers code.

**Comment by Grokzen** [comment]
Also, could you add a line to the CHANGES file with this major change? Could you add some minor text like how much it could improve performance etc.

**Comment by Grokzen** [comment]
@72squared I did some minor tweaks in `docs/pipeline.md` so you will have to rebase this PR on the latest code in unstable and fix the conflicts before i can merge this.

File: rediscluster/pipeline.py [hunk 4]
```diff
@@ -534,22 +369,45 @@ def inner(*args, **kwargs):
...
+    def append(self, c):
+        self.commands.append(c)
 
-    def run(self):
+    def write(self):
+        """
+        Code borrowed from StrictRedis so it can be fixed
+        """
+        # build up all commands into a single request to increase network perf
+        connection = self.connection
```
Comment (Grokzen) [inline rediscluster/pipeline.py]: Do we need local tmp variable here?

File: rediscluster/pipeline.py [hunk 5]
```diff
@@ -534,22 +369,45 @@ def inner(*args, **kwargs):
...
-            self.value = self.execute(self.conn, self.cmds, False)
-        except Exception as e:
-            self.exception = e
+            connection.send_packed_command(connection.pack_commands([c.args for c in commands]))
+        except ConnectionError as e:
+            for c in commands:
+                c.result = e
+
+    def read(self):
+        connection = self.connection
```
Comment (Grokzen) [inline rediscluster/pipeline.py]: Same here with local tmp variable

Comment (72squared) [inline rediscluster/pipeline.py]: I did it as an optimization for the loop iteration. local variables are faster to access. But really it probably doesn't matter. Do you have a strong preference?

Comment (Grokzen) [inline rediscluster/pipeline.py]: Hmm. Keep it as is.

**Comment by Grokzen** [comment]
LGTM :]

**Pull request merged**